File: Leetcode/78. Subsets.cpp
class Solution {
public:
    vector<int>a;
    vector<vector<int>>ans;

    void f(vector<int>b, int ind) {
        if (ind == a.size()) {
            ans.push_back(b);
            return;
        }
        f(b, ind + 1);
        b.push_back(a[ind]);
        f(b, ind + 1);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        a = nums;
        f({}, 0);
        return ans;
    }
};

File: Leetcode/78. Subsets.java
class Solution {
    private void f(int[] nums, List<Integer>cur, int ind, List<List<Integer>>ans) {
        if (ind == nums.length) {
            ans.add(new ArrayList<>(cur));
            return;
        }
        f(nums, cur, ind + 1, ans);
        cur.add(nums[ind]);
        f(nums, cur, ind + 1, ans);
        cur.remove(cur.size() - 1);
    }

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>>ans = new ArrayList<>();
        List<Integer>cur = new ArrayList<>();
        f(nums, cur, 0, ans);
        return ans;
    }
}
